doc/adr/0017-state-machines-for-resource-state-management.md

```diff
@@ -0,0 +1,47 @@
+# 0017 State machines for resource state management
+
+Date: Unknown
+
+## Status
+
+Accepted
+
+## Context
+
+The application consists of many, wildly different, types of resources, such
+as invoices, orders, e-mails, document templates, recurrence patterns, etc.
+Each of these resources has a set of attributes which together imply the state
+of the resource *and* the allowable actions to be invoked on it.  As a result,
+the developer needs a high level of knowledge of the resource, its implied state
+and the allowable invokable actions to keep the application state consistent.
+
+By making both the state of the resource and the invokable actions (state
+transitions) explicit, the application provides documentation to less
+experienced developers.  Modelling state machines challenges developers to
+step out of the context of the code they are writing, considering more broadly
+the impact of any planned changes.  Additionally, the application gains built-in
+controls to assure its internal state to remain consistent.
+
+Even more: if the state machine allows for introspection, allowing the
+developer to query the available list of invokable actions, the state machine
+could be used to guide the user through the process (without hard-coding it into
+the client).
+
+
+## Decision
+
+Considering that LedgerSMB is an accounting application and that accounting needs
+the most stringent controls on consistency of the internal state available, state
+machines will be used to keep track of internal resources such as invoices,
+gl transactions and e-mails.
+
+## Consequences
+
+1. A state machine engine must be selected
+2. Existing resources must be enhanced with state machines
+   and their implied state must be reverse-engineered
+3. New resources must not be introduced without state machines
+4. A design needs to be created mapping available state machine
+   actions to web-API responses and requests to trigger actions
+
```

Comment:
5. State transitions must be logged, with the time of the transition and the user initiating it.

(suggestion... otherwise this looks good).

This does make me think of another ADR to consider -- move the permission model from applying only to data access, to one that is used to enforce access on state transitions.

There are probably a few places where we want to restrict all access to particular columns/rows in the database -- but for the most part this is way too granular, and workflow transitions are a lot more appropriate place to enforce access controls.

Reply:
> This does make me think of another ADR to consider -- move the permission model from applying only to data access, to one that is used to enforce access on state transitions.

This is (at a very small scale) already what I'm doing: e.g. posting a transaction or invoice is restricted by ACL (https://github.com/ledgersmb/LedgerSMB/blob/master/lib/LedgerSMB/Workflow/Condition/ACL.pm) where a user is being checked to have the right role to even be offered the button (but enforcement is within the workflow, not at the UI level).

Reply:
So, yes, lets write that ADR too.

Reply:
Suggestion to this extent added to the README which lists the "ADRs still to be done"